packages/flutter/test/widgets/framework_test.dart

```diff
@@ -61,6 +61,199 @@ void main() {
+            _Stateful(
+              child: Text(
+                'Text2',
+                textDirection: TextDirection.ltr,
+                key: key2,
+              ),
+            ),
+          ],
+        ),
+      )
+    );
+    final _StatefulState state = tester.firstState(find.byType(_Stateful).at(1));
+    state.rebuild();
+    // Reorder the items
+
+    await tester.pumpWidget(
+      LayoutBuilder(
+        builder: (BuildContext context, BoxConstraints constraints) => Column(
+          children: const <Widget>[
+            _Stateful(
+              child: Text(
+                'Text2',
+                textDirection: TextDirection.ltr,
+                key: key2,
+              ),
+            ),
+            _Stateful(
+              child: Text(
+                'Text1',
+                textDirection: TextDirection.ltr,
+                key: key1,
+              ),
+            ),
+          ],
+        ),
+      )
+    );
+    expect(tester.takeException(), null);
+  });
+
+  testWidgets('GlobalKey correct case 4 - can deal with early rebuild in layoutbuilder - move forward', (WidgetTester tester) async {
+    const Key key1 = GlobalObjectKey('Text1');
+    const Key key2 = GlobalObjectKey('Text2');
+    const Key key3 = GlobalObjectKey('Text3');
+    await tester.pumpWidget(
+      LayoutBuilder(
+        builder: (BuildContext context, BoxConstraints constraints) => Column(
+          children: const <Widget>[
+            _Stateful(
+              child: Text(
+                'Text1',
+                textDirection: TextDirection.ltr,
+                key: key1,
+              ),
+            ),
+            _Stateful(
+              child: Text(
+                'Text2',
+                textDirection: TextDirection.ltr,
+                key: key2,
+              ),
+            ),
+            _Stateful(
+              child: Text(
+                'Text3',
+                textDirection: TextDirection.ltr,
+                key: key3,
+              ),
+            ),
+          ],
+        ),
+      )
+    );
+    final _StatefulState state = tester.firstState(find.byType(_Stateful).at(1));
+    state.rebuild();
+    // Reorder the items
+
+    await tester.pumpWidget(
+      LayoutBuilder(
+        builder: (BuildContext context, BoxConstraints constraints) => Column(
+          children: const <Widget>[
+            _Stateful(
+              child: Text(
+                'Text1',
+                textDirection: TextDirection.ltr,
+                key: key1,
+              ),
+            ),
+            _Stateful(
+              child: Text(
+                'Text3',
+                textDirection: TextDirection.ltr,
+                key: key3,
+              ),
+            ),
+            _Stateful(
+              child: Text(
+                'Text2',
+                textDirection: TextDirection.ltr,
+                key: key2,
+              ),
+            ),
+          ],
+        ),
+      )
+    );
```

Comment:
ditto

Reply:
(checking that the rebuild happens before the relayout -- looking at this closer, i'm not sure it actually is)

Reply:
I don't think it can happen after relayout, rebuild happens in build phase while relayout happens in layout phase. I will try to come up with the check anyway though.

Reply:
there's actually no such thing as a build phase and a layout phase. It's a white lie we tell people to give them an intuition of how things work. The reality is that build and layout are interleaved. That's actually the whole point of a LayoutBuilder. It builds, then stops the build at that point, waits until it's laid out, stops the layout at that point, builds again, then lays out.

Reply:
yes, you are right. I added the check to make sure the build order matches expectation